### PR TITLE
Implement missing optional attributes for general diagnostics cluster

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
What is being fixed?  Examples:
* We have three left optional attributes which need implementation in general diagnostic cluster 

#### Change overview
Implement missing optional attributes for general diagnostic cluster

#### Testing
How was this tested? (at least one bullet point required)
```
yufengw@yufengw-SEi:~/connectedhomeip/out/debug/standalone$ ./chip-tool generaldiagnostics read active-hardware-faults 12344321 0
[1637006901.600958][50024:50029] CHIP:DMG: ReportDataMessage =
[1637006901.600962][50024:50029] CHIP:DMG: {
[1637006901.600966][50024:50029] CHIP:DMG: 	AttributeReportIBs =
[1637006901.600970][50024:50029] CHIP:DMG: 	[
[1637006901.600973][50024:50029] CHIP:DMG: 		AttributeReportIB =
[1637006901.600977][50024:50029] CHIP:DMG: 		{
[1637006901.600981][50024:50029] CHIP:DMG: 			AttributeDataIB =
[1637006901.600985][50024:50029] CHIP:DMG: 			{
[1637006901.600989][50024:50029] CHIP:DMG: 				AttributePathIB =
[1637006901.600996][50024:50029] CHIP:DMG: 				{
[1637006901.601001][50024:50029] CHIP:DMG: 					Endpoint = 0x0,
[1637006901.601006][50024:50029] CHIP:DMG: 					Cluster = 0x33,
[1637006901.601011][50024:50029] CHIP:DMG: 					Attribute = 0x0000_0005,
[1637006901.601016][50024:50029] CHIP:DMG: 				}
[1637006901.601021][50024:50029] CHIP:DMG: 					
[1637006901.601026][50024:50029] CHIP:DMG: 					Data = [
[1637006901.601032][50024:50029] CHIP:DMG: 						1, 2, 5, 8, 
[1637006901.601037][50024:50029] CHIP:DMG: 					],
[1637006901.601041][50024:50029] CHIP:DMG: 				DataVersion = 0x0,
[1637006901.601045][50024:50029] CHIP:DMG: 			},
[1637006901.601051][50024:50029] CHIP:DMG: 			
[1637006901.601054][50024:50029] CHIP:DMG: 		},
[1637006901.601059][50024:50029] CHIP:DMG: 		
[1637006901.601063][50024:50029] CHIP:DMG: 	],
[1637006901.601070][50024:50029] CHIP:DMG: 	
[1637006901.601073][50024:50029] CHIP:DMG: }
[1637006901.601099][50024:50029] CHIP:ZCL: ReadAttributesResponse:
[1637006901.601102][50024:50029] CHIP:ZCL:   ClusterId: 0x0000_0033
[1637006901.601106][50024:50029] CHIP:ZCL:   attributeId: 0x0000_0005
[1637006901.601110][50024:50029] CHIP:ZCL:   status: Success                (0x0000)
[1637006901.601114][50024:50029] CHIP:ZCL:   attribute TLV Type: 0x16
[1637006901.601120][50024:50029] CHIP:TOO: OnGeneralDiagnosticsActiveHardwareFaultsListAttributeResponse: 4 entries
[1637006901.601128][50024:50029] CHIP:TOO:   [1]: 1
[1637006901.601132][50024:50029] CHIP:TOO:   [2]: 2
[1637006901.601136][50024:50029] CHIP:TOO:   [3]: 5
[1637006901.601139][50024:50029] CHIP:TOO:   [4]: 8
```

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

File: src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp [hunk 1]
```diff
@@ -67,6 +72,32 @@ CHIP_ERROR GeneralDiagosticsAttrAccess::ReadIfSupported(CHIP_ERROR (PlatformMana
...
+            {
+                ReturnErrorOnFailure(encoder.Encode(fault));
+            }
+
+            return CHIP_NO_ERROR;
+        });
+    }
+    else
+    {
+        err = aEncoder.Encode(DataModel::List<EndpointId>());
```
Comment (Damian-Nordic) [inline src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp:95]: Why do we encode a list of `EndpointId`s?

File: src/include/platform/GeneralFaults.h [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+    int size() const;
+    uint8_t operator[](int index) const;
+
+    Iterator begin() const;
+    Iterator end() const;
+
+private:
+    uint8_t mData[N];
+    int mSize = 0;
```
Comment (Damian-Nordic) [inline src/include/platform/GeneralFaults.h:66]: Use `size_t`? Or `uint8_t` to reduce the memory usage.

File: src/include/platform/GeneralFaults.h [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+inline uint8_t GeneralFaults<N>::operator[](int index) const
+{
+    VerifyOrDie(index < mSize);
+    return mData[index];
+}
+
+template <size_t N>
+inline typename GeneralFaults<N>::Iterator GeneralFaults<N>::begin() const
+{
+    return GeneralFaults<N>::Iterator{ this, 0 };
```
Comment (Damian-Nordic) [inline src/include/platform/GeneralFaults.h:102]: You don't need a specialized iterator. You can define `iterator` as `uint8_t*` and return `mData` here. Likewise, `end()` may return `&mData[mSize]`.

File: src/platform/Linux/PlatformManagerImpl.cpp [hunk 4]
```diff
@@ -402,6 +402,38 @@ CHIP_ERROR PlatformManagerImpl::_GetBootReasons(uint8_t & bootReasons)
     return err;
 }
 
+CHIP_ERROR PlatformManagerImpl::_GetActiveHardwareFaults(GeneralFaults<kMaxHardwareFaults> & hardwareFaults)
+{
+    // On Linux Simulation, set following hardware faults statically.
```
Comment (Damian-Nordic) [inline src/platform/Linux/PlatformManagerImpl.cpp:407]: This is not only used in tests, so I think it should be guarded with some `#if` to only enable it for testing.

**Review by Damian-Nordic — CHANGES_REQUESTED** [review]

File: src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp (hunk 1 above)
Comment (yufengwangca) [reply src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp:95]: To pas CI check, all attributes enabled in controller clusters are tested via a read in the darwin tests, It's just a simple success/failure test.

File: src/include/platform/GeneralFaults.h (hunk 3 above)
Comment (yufengwangca) [reply src/include/platform/GeneralFaults.h:102]: You mean I can just replace following Iterator define with` typedef uint8_t * Iterator;` ?
```
    /* The iterator */
    class Iterator
    {
    public:
        Iterator(const GeneralFaults<N> * GeneralFaults, int index);
        uint8_t operator*() const;
        Iterator & operator++();
        bool operator!=(const Iterator & other) const;

    private:
        const GeneralFaults<N> * mGeneralFaultsPtr;
        int mIndex = -1;
    };
```

Comment (Damian-Nordic) [reply src/include/platform/GeneralFaults.h:102]: Yes, see for example how `Span` implements `begin()` and `end()`. Returning pointers is enough to use a container in `for (auto a: container)` loop. :)

**Review by Damian-Nordic — APPROVED** [review]
Reworking the iterator is optional, so approving.

File: src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp (hunk 1 above)
Comment (yufengwangca) [reply src/app/clusters/general_diagnostics_server/general_diagnostics_server.cpp]: good catch, it should be unit8_t. Thanks!

File: src/include/platform/GeneralFaults.h (hunk 2 above)
Comment (yufengwangca) [reply src/include/platform/GeneralFaults.h:66]: we need to use -1 to indicate invalid index

File: src/include/platform/PlatformManager.h [hunk 5]
```diff
@@ -495,5 +500,20 @@ inline CHIP_ERROR PlatformManager::GetBootReasons(uint8_t & bootReasons)
...
+inline CHIP_ERROR PlatformManager::GetActiveRadioFaults(GeneralFaults<kMaxRadioFaults> & radioFaults)
+{
+    return static_cast<ImplClass *>(this)->_GetActiveRadioFaults(radioFaults);
+}
+
+inline CHIP_ERROR PlatformManager::GetActiveNetworkFaults(GeneralFaults<kMaxNetworkFaults> & networkFaults)
+{
+    return static_cast<ImplClass *>(this)->_GetActiveNetworkFaults(networkFaults);
+}
+
```
Comment (tcarmelveilleux) [inline src/include/platform/PlatformManager.h:517]: This should be from a new interface. We are adding too much to the PlatformManager.

Furthermore, this getter interface prevents supporting events about those faults.

Comment (yufengwangca) [reply src/include/platform/PlatformManager.h:517]: Adding a new platform dependent interface is quite expensive, including adding a new manager and generic implementation and all platforms need to implement the interface to override the default behaviors.

**Review by bzbarsky-apple — APPROVED** [review]
Approving so this is unblocked once @tcarmelveilleux's issues are resolved.

File: src/include/platform/PlatformManager.h (hunk 5 above)
Comment (yufengwangca) [reply src/include/platform/PlatformManager.h:517]: Per sync offline, we will introduce DiagnosticDataProvider + DiagnosticsDelegate for all diagnostics clusters. Only handle these three attributes within this PR with new interface and will refactor all other diagnostics attributes to use this new interface in the followup PRs

Comment (yufengwangca) [reply src/include/platform/PlatformManager.h:517]: This PR has been merged, track the followup work at https://github.com/project-chip/connectedhomeip/issues/11922

**Pull request merged**